File: shoring/desk/src/main/java/com/panopset/marin/fx/ImageMapCache.java
package com.panopset.marin.fx;

import java.util.HashMap;
import java.util.Map;
import com.panopset.compat.Logop;
import javafx.scene.image.Image;

public enum ImageMapCache {

  INSTANCE;

  private ImageMapCache() {}

  public static Image get(String key) {
    return INSTANCE.map.get(key);
  }

  public static Image get(String key, String path) {
    return INSTANCE.find(key, path);
  }

  Image find(String key, String path) {
    Image rtn = map.get(key);
    if (rtn == null) {
      try {
        rtn = new Image(getClass().getResource(path).toExternalForm());
      } catch (IllegalArgumentException iae) {
        Logop.error(path, iae);
        return null;
      }
      map.put(key, rtn);
    }
    return rtn;
  }

  private Map<String, Image> map = new HashMap<>();
}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/fx/ImageMapCache.kt
package com.panopset.marin.fx

import com.panopset.compat.Logz
import javafx.scene.image.Image
import javafx.scene.image.WritableImage

object ImageMapCache {

    private fun find(key: String, path: String): Image {
        var rtn = map[key]
        if (rtn == null) {
            rtn = try {
                Image(javaClass.getResource(path)?.toExternalForm())
            } catch (iae: IllegalArgumentException) {
                Logz.errorMsg(path, iae)
                WritableImage(0, 0)
            }
            map[key] = rtn?: WritableImage(0, 0)
        }
        return rtn?: WritableImage(0, 0)
    }

    private val map: MutableMap<String, Image> = HashMap()

    operator fun get(key: String): Image? {
        return map[key]
    }

    operator fun get(key: String, path: String?): Image? {
        if (path == null) {
            return null
        }
        return find( key, path)
    }
}
